pkg/scheduler/objects/application.go

```diff
@@ -886,6 +886,40 @@ func (sa *Application) tryAllocate(headRoom *resources.Resource, nodeIterator fu
 					zap.String("required node", requiredNode))
 				return nil
 			}
+
+			// Are there any non daemon set reservations on specific required node?
+			// Cancel those reservations to run daemon set pods
+			reservations := node.getReservations()
+			if len(reservations) > 0 {
+				unreserve := true
+				for _, res := range node.getReservations() {
+					// skip the node
+					if res.ask.GetRequiredNode() != "" {
+						unreserve = false
+						break
+					}
+				}
+				if unreserve {
+					// un reserve all the apps that were reserved on the node
+					reservedKeys, releasedAsks := node.UnReserveApps()
+					if len(reservedKeys) == 0 || len(releasedAsks) == 0 {
```

Comment:
The comments on the `node.UnReserveApps()` method are wide ranging: we do not lock the node or the app inside this method. It also does a callback into the application which is unlocked. This can cause data races. We *must* not use this method.

The implementation should, if we do not skip the node, iterate over the list of reservations returned and call `app.UnReserve(node, ask)` to make sure we do the correct locking etc.
Logging is simply handed for each returning call.